[Image Resizer] Fix image resizer unexpected property type or value (#9885, #14266)

CleanCodeDeveloper opened the pull request:
## Summary of the Pull Request

**What is this about:**

Improves resilience of the image resize implementation.

This PR fixes different categories of bugs in described in #9885 and #14266.

Bug category 1:
- Image contains partially invalid metadata which throws an exception on:
  - reading a given metadata field
  - writing metadata to a given field or persisting the image containing the metadata to disk. 

Bug category 2:
- Some metadata depends on the availability of a thumbnail image. The encoder crashes if the thumbnail image is not found.

**What is include in the PR:** 
- Loop over all metadata and try to read the actual value. Remove metadata item if exception is thrown while reading the value. Build up the metadata object from the scratch by adding all valid values. This resolves Bug category 1
- copy thumbnail image from input to output image to resolve Bug category 2
- Adds two metadata debug methods to simplify the debugging of metadata related bugs 

**How does someone test / validate:** 
test with images provided in linked issue.

Images: Bug category 1:
[#9885 (comment)](https://github.com/microsoft/PowerToys/issues/9885#issuecomment-958967927)
[#9885 (comment)](https://github.com/microsoft/PowerToys/issues/9885#issuecomment-891829849)
[#14266 (comment)](https://github.com/microsoft/PowerToys/issues/14266#issue-1044400800)

Images: Bug category 2:
[#9885 (comment)](https://github.com/microsoft/PowerToys/issues/9885#issue-815762941)

## Quality Checklist

- [x] **Linked issue:** #9885, #14266
- [x] **Communication:** I've discussed this with core contributors in the issue. 
- [ ] **Tests:** Added/updated and all pass
- [ ] **Installer:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Docs:** Added/ updated
- [ ] **Binaries:** Any new files are added to WXS / YML
   - [ ] No new binaries
   - [ ] [YML for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.

htcfreek commented:
I think we should log a warning if we remove corrupt metadata while resizing. This helps us on incomming issues because the normal logs aren't debug logs.

I suggest something like this: `[Warn] Removing corrupt metadata property <name> from inage <path>. - Reason: <reason of corruption>`

CleanCodeDeveloper commented:
> First of all thanks for your contribution!
> 
> **My thoughts on your code:** I think we should log a warning if we remove corrupt metadata while resizing. This helps us on incomming issues because the normal logs aren't debug logs.
> 
> I suggest something like this:
> 
> ```
> [Warn] Removing corrupt metadata property <name> from image <path>. - Reason: <reason of corruption>
> ```

Thank you for your suggestion. I agree that improved logging would be a good thing!

There is currently no logger within the Image Resizer project. Other modules like FancyZone do have a [logger implementation](https://github.com/microsoft/PowerToys/blob/master/src/modules/fancyzones/editor/FancyZonesEditor/Logs/Logger.cs). I am wondering if adding a Logger would be beyond the scope of this PR.

For now, I added a few debug logs because of sanity. Those can be easily converted once a "real" logger is available.

htcfreek commented:
> > First of all thanks for your contribution!
> > 
> > **My thoughts on your code:** I think we should log a warning if we remove corrupt metadata while resizing. This helps us on incomming issues because the normal logs aren't debug logs.
> > 
> > I suggest something like this:
> > 
> > ```
> > [Warn] Removing corrupt metadata property <name> from image <path>. - Reason: <reason of corruption>
> > ```
> 
> Thank you for your suggestion. I agree that improved logging would be a good thing!
> 
> There is currently no logger within the Image Resizer project. Other modules like FancyZone do have a [logger implementation](https://github.com/microsoft/PowerToys/blob/master/src/modules/fancyzones/editor/FancyZonesEditor/Logs/Logger.cs). I am wondering if adding a Logger would be beyond the scope of this PR.
> 
> For now, I added a few debug logs because of sanity. Those can be easily converted once a "real" logger is available.

There is an open PR for implement logging.

htcfreek commented:
@stefansjfw , @SeraphimaZ
Can we use the common logger from FZ and PowerRename here?

stefansjfw commented:
> @stefansjfw , @SeraphimaZ Can we use the common logger from FZ and PowerRename here?

Not out-of-the-box easily. Let's do it as a follow-up. no need to block this

htcfreek commented:
@stefansjfw , @CleanCodeDeveloper 
Let's add comments with the log text where we want to log things as placeholder for future changes.

CleanCodeDeveloper commented:
> @stefansjfw , @CleanCodeDeveloper Let's add comments with the log text where we want to log things as placeholder for future changes.

I added a few `Debug.Writeline` statements instead of comment log statements because they already provide valuable output during debugging. Those debug statements come at no cost as there are not included in the release binary.

> If you use methods in the Debug class to print debugging information and check your logic with assertions, you can make your code more robust without impacting the performance and code size of your shipping product.
[Source: Microsoft Docs, System.Diagnostics.Debug](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debug?view=net-6.0#remarks)

As soon as we have a proper logger implementation we can replace
```
Debug.WriteLine($"Exception while trying to set metadata {value} at position: {query}");
Debug.WriteLine(ex);
```
by something like
```
_logger.LogError(ex, $"Exception while trying to set metadata {value} at position: {query}");
```

stefansjfw commented on src/modules/imageresizer/ui/Models/ResizeOperation.cs:
```diff
@@ -105,9 +131,9 @@ public void Execute()
+                            originalFrame.Thumbnail,
                             metadata,
-                            colorContexts: null));
+                            originalFrame.ColorContexts));
```
If I'm not wrong, only these 2 lines change also fixes the issue. can someone confirm? If yes, we don't need to modify metadata

stefansjfw replied:
Lines 108 and 110 originally

CleanCodeDeveloper replied:
You are right, this fixes the initial issue described in #9885. I refer to this as "Bug category 2" in the PR description.

The rest of the code is required to fix the "Bug category 1". Please find a few sample images here:
[#9885 (comment)](https://github.com/microsoft/PowerToys/issues/9885#issuecomment-958967927)
[#9885 (comment)](https://github.com/microsoft/PowerToys/issues/9885#issuecomment-891829849)
[#14266 (comment)](https://github.com/microsoft/PowerToys/issues/14266#issue-1044400800)

CleanCodeDeveloper replied:
Btw: thank you for taking the time to review my PR!

stefansjfw reviewed (approved):
Looks good to me. I'm not the expert when it comes to image files and metadata, though.. Hopefully, eventually removed metadata is not the most important ones and won't hurt Image Resizer much :)

crutkas commented:
@jaimecbernardo what is needed to merge in?

CleanCodeDeveloper commented:
> Hopefully, eventually removed metadata is not the most important ones and won't hurt Image Resizer much :)

It only removes metadata which causes an exception on read/write so this should be fine! 

I will asked @ChrisJohnTaylor to perform a user acceptance test once this is (pre-)released.

crutkas commented:
@CleanCodeDeveloper / @jaimecbernardo should we create new issues for the type of stuff we haven't resolved here?  Right now i'm trying to understand what is and is not fix and honestly, i think it may be a good idea to call them done and xref new issues with what is not fixed.

Thoughts?

CleanCodeDeveloper commented:
This PR should fix all reported metadata issues. I checked it with all images in issue #9885 and comments. It also fixes #14266.

I searched for open issues about problems to resize certain images but could not find any. Please let me know if there are any.

jaimecbernardo commented:
> @CleanCodeDeveloper / @jaimecbernardo should we create new issues for the type of stuff we haven't resolved here? Right now i'm trying to understand what is and is not fix and honestly, i think it may be a good idea to call them done and xref new issues with what is not fixed.
> 
> Thoughts?

It's not clear what kind of metadata fields might cause issues for the Encoder. I think these will have to be added as they're detected by users.

jaimecbernardo reviewed (changes requested):
Can this be made into an option, as suggested [in the original proposal](https://github.com/microsoft/PowerToys/issues/9885#issuecomment-969355640)?
"Remove metadata which causes problems in some cases"

The main reason for this is that people might still find the metadata we are removing useful in other scenarios. For those cases, the `_ = metadata.Clone();` could still be taken.

The adding a thumbnail change is good and can stay even if the option is not selected.

Other than that, it looks great to me!

stefansjfw commented:
> Can this be made into an option, as suggested [in the original proposal](https://github.com/microsoft/PowerToys/issues/9885#issuecomment-969355640)? "Remove metadata which causes problems in some cases"
> 
> The main reason for this is that people might still find the metadata we are removing useful in other scenarios. For those cases, the `_ = metadata.Clone();` could still be taken.
> 
> The adding a thumbnail change is good and can stay even if the option is not selected.
> 
> Other than that, it looks great to me!

We already have "Remove metadata that doesn't affect rendering option" which make ImageResizer work for images mentioned in this issue.. Do we want to have both 
- "Remove metadata that doesn't affect rendering option"
- "Remove metadata which causes problems in some cases" ?

jaimecbernardo commented:
> > Can this be made into an option, as suggested [in the original proposal](https://github.com/microsoft/PowerToys/issues/9885#issuecomment-969355640)? "Remove metadata which causes problems in some cases"
> > The main reason for this is that people might still find the metadata we are removing useful in other scenarios. For those cases, the `_ = metadata.Clone();` could still be taken.
> > The adding a thumbnail change is good and can stay even if the option is not selected.
> > Other than that, it looks great to me!
> 
> We already have "Remove metadata that doesn't affect rendering option" which make ImageResizer work for images mentioned in this issue.. Do we want to have both
> 
> * "Remove metadata that doesn't affect rendering option"
> * "Remove metadata which causes problems in some cases" ?

In the issue some users mentioned that they still wanted some metadata regarding camera settings, while not wanting to hit these errors for those images.

ChrisJohnTaylor commented:
As the one who started the original thread https://github.com/microsoft/PowerToys/issues/9885#issuecomment-890903077  back in February, I can say I would not use Image Resizer (fantasic and easy-to-use as it is!) if it removed EXIF tags that are important to photographers referring back to what was going on with their images. In particular, tags dealing with make model of the camera and lens, orientation, exposure time, FNumber, ExposureProgram, ISOSpeedRatings, DateTimeOriginal, OffsetTime, ExposureBiasValue, MeteringMode, FocalLength, CoilorSpace, Flash, any GPS info. Another few that I believe are used by some programs to decide how to display or print the image are XResolution, YResolution, and ResolutionUnit. I don't know what a lot of the other tags are used for, so I don't know if they are important or not for some obscure reason.

Going back to the other thread,

**I am not a coder!!**, so I don't know if this is practical or possible or too much work. I wonder if an alternative might be to use the code within - https://sourceforge.net/projects/jpegview/ that writes the metadata. It seems to have no problems handling the metadata in the problem images I have (which start as Nikon NEF RAW files and were saved as JPG files by various programs)

My problem images all started as Nikon RAW (.NEF) files. (but others have indicated it has been a problem with their non-Nikon files. serafisnz mentioned in the other thread that he uses a Sony camera.)

- ON1 Photo RAW exports a JPG but the metadata in the resulting export still causes Image Resizer to choke
- IrfanView exports a JPG but the metadata in the resulting export still causes Image Resizer to choke
- PhotoScape X exports a JPG ... AND seems to fix the metadata - Image Resizer can process it fine

So other programs are managing to write JPG files even if the property values are "bad". ON1 Photo RAW, Irfanview, and PhotoScape X are not open source so it is not possible to see how they are doing it, but JPEGVIEW can write JPG files that have such "bad" EXIF data and it is open source. So if it would be possible to use whatever technique they are using to write EXIF data, I think the issue here could be considered solved.

Otherwise, if it is not possible to change the code that writes the EXIF info, I think it is really important to preserve as much "good" EXIF info as possible.

Thanks.

CleanCodeDeveloper commented:
@jaimecbernardo: I deviated from the original proposal because I found that we need a more generic approach.

Let me share my thought process that you could better understand why I choose to implement it as it is. Sorry about not making this clear.

I started with the idea to add an UI toggle "Remove metadata which causes problems in some cases” and remove  metadata fields which contained malformed metadata in the past as described [here](https://github.com/microsoft/PowerToys/issues/9885#issuecomment-969355640). 

I analyzed the first image, identified the problematic metadata field, and removed it. I thought I would be done but discovered that other images in the comments of #9885 had different invalid metadata fields. So, I added more fields to my list of problematic metadata.

The problem is that in theory each field could contain malformed metadata. I needed a way to verify that I only remove the given field when its malformed. I discovered that `BitmapMetadata` throws an exception on `GetQuery` when you read an malformed metadata field.

It took me way longer than it should to realize that `BitmapMetadata` implements `IEnumerable` and that I could enumerate over all metadata. 
I altered my approach to generate the list of problematic metadata fields by iteration over all metadata and remove all fields where I get an exception while accessing a field with `GetQuery`.

I was convinced again that I solved everything. Unfortunately, I still got some errors with some images (exception on `_encoder.Save(…)`). I discovered that some metadata fields can be read without problems but will throw an exception one write either with `SetQuery` or ultimately while executing `_encoder.Save(…)`.

It became clear to me that the only approach is to rebuild the metadata object from the scratch and ignore all fields which throw an exception in either `GetQuery` or ` SetQuery`.

My final solution looks like this:

```cs
// read all metadata and build up metadata object from the scratch. Discard invalid (unreadable/unwritable) metadata.
// create new root metadata object
var newMetadata = new BitmapMetadata(metadata.Format);
// store temporary list of metadata excluding metadata fields which throw an exception on read (GetQuery)
var listOfMetadata = modifiableMetadata.GetListOfMetadata();
// iterate over all metadata and add it to the new metadata root
foreach (var (metadataPath, value) in listOfMetadata)
{
  // metadata is a tree like data structure. Determine if it’s a node and not a leaf
  if (value is BitmapMetadata bitmapMetadata)
  {
    // create new metadata node
    var innerMetadata = new BitmapMetadata(bitmapMetadata.Format);

    // add the node to the metadata root object 
    newMetadata.SetQuerySafe(metadataPath, innerMetadata);
  }
  else
  {
    // it’s a leaf, add to metadata object. Ignore metadata which causes an exception in write (SetQuery)
    newMetadata.SetQuerySafe(metadataPath, value);
   }
 }
// replaces the old metadata root object with the newly created and validated metadata object
metadata = newMetadata;
```

To wrap things up:
-	We don’t remove any metadata for most of the images. We only remove metadata which will cause an exception on encoder.Save(...)
- 	The approach is very generic and does not require manual maintenance of an ever-growing list of metadata fields which where problematic in the past

jaimecbernardo commented:
@CleanCodeDeveloper , got it! Thank you for the explanation. This is a superior solution, indeed! Great work.

jaimecbernardo reviewed (approved):
LGTM! Great work.
Thank you for the contribution!

ChrisJohnTaylor commented:
@CleanCodeDeveloper - Thanks for the explanation. this certainly sounds like it will fix things once and for all! It was suggested at some point that I could be provided with a pre-release of the final program and I could test on as wide a variety of images as I can. I would be happy to do so. Please keep in mind I am not able to build things from source or anything :-) I need a simple installation or file replacement.

Thanks again for all your work on this.

CleanCodeDeveloper commented:
@ChrisJohnTaylor: I think release of version 0.51 is scheduled for end of November. Not sure if we get a pre-release this time. Lets shift our discussion over to #9885 that we don't annoy the maintainers with unnecessary notifications ;-)

crutkas commented:
@jaimecbernardo is this in 51 or 53?  
@CleanCodeDeveloper  does this include #14266 as well as #9885?

jaimecbernardo commented:
It's in .51

CleanCodeDeveloper commented:
> @CleanCodeDeveloper does this include #14266 as well as #9885?

Yes, this PR fixes both #14266 and #9885

bricelam commented on src/modules/imageresizer/ui/Models/ResizeOperation.cs:
```diff
@@ -105,9 +131,9 @@ public void Execute()
+                            originalFrame.Thumbnail,
                             metadata,
-                            colorContexts: null));
+                            originalFrame.ColorContexts));
```
Interesting, I took this out a few years ago because it was screwing up the color on some images. See https://github.com/bricelam/ImageResizer/issues/53

bricelam replied:
Ah, looks like you've since taken it out again :-)

CleanCodeDeveloper replied:
yeah, stupid me added it because it looked "right". I reverted this change with #14867. Btw: I'm impressed that you remember this issue after all this years
